nixos/pipewire: add transparent ALSA/PulseAudio/JACK support

nglen opened the pull request:
###### Motivation for this change
This allows ALSA programs to interact through PipeWire.

The options are the same as in the JACK module for consistency. I enabled it by default since that is what JACK does and it seems less confusing for end users however this technically makes it backwards incompatible. If you prefer it off by default I can do that instead.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

nglen committed:
nixos/pipewire: add ALSA routing

This code is based on the similar implementation for JACK.

nglen committed:
nixos/pipewire: add jack/pulse emulation

This allows for transparent JACK and PulseAudio emulation. With this you
can essentially replace your entire audio framework with just PipeWire
for almost no configuration.

nglen commented:
I just added both JACK and PulseAudio transparent emulation support. This allows running dependent programs using PipeWire as a backend instead. I also decided to set the ALSA support to false by default which not only is consistent with these new options but also makes it fully backwards compatible now.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/244

bbigras commented:
If I use `services.pipewire.enable = true;` without enabling `services.pipewire.alsa.enable = true;` I get:

```
error: The option `nodes.desktop.configuration.environment.etc.alsa/conf.d/50-pipewire.conf.source' is used but not defined.
(use '--show-trace' to show detailed location information)
```

ashkitten commented:
can you give more info for how to test this? i was not able to get any kind of audio output with any of `aplay`, `paplay`, or `jack_simple_session` with this. `qjackctl` saw a jack server running, but `jack_simple_session` was not able to connect to it and neither was `cadence`. `pavucontrol` listed a number of devices in the "configuration" tab, but the only setting for any of them was "off".

bbigras commented:
Similar thing to me. Devices are off.

pipewire log:
```
-- Logs begin at Mon 2020-05-18 13:42:44 EDT, end at Tue 2020-08-04 14:37:37 EDT. --
aoû 04 14:34:43 desktop systemd[1811]: Started Multimedia Service.
aoû 04 14:34:48 desktop pipewire[2016]: [E][000000038.312224][bluez5-dbus.c:1436 get_managed_objects_reply()] GetManagedObjects() failed: org.freedesktop.DBus.Error.TimedOut
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.395758][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.401733][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.404457][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.407094][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.415894][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.418918][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.421218][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:0 seq:130 res:-22 (Invalid argument) msg:"unknown resource 26 op:1"
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.422063][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.425590][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.428402][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.433018][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:0 seq:144 res:-22 (Invalid argument) msg:"unknown resource 26 op:1"
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.434648][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.436970][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.439274][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.441689][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.444313][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.446197][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:26 seq:146 res:-2 (No such file or directory) msg:"no global 41"
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.446211][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:0 seq:148 res:-22 (Invalid argument) msg:"unknown resource 26 op:7"
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.453788][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.458266][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.460643][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.463019][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.465516][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.467776][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.470177][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.472528][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.474867][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.477521][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.480291][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.483935][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.487254][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.489755][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2016]: [W][000000045.492359][policy-node.c:549 rescan_node()] no node found for 41
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.502975][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:26 seq:166 res:-2 (No such file or directory) msg:"no global 41"
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.503020][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:0 seq:168 res:-22 (Invalid argument) msg:"unknown resource 26 op:7"
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.503026][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:30 seq:169 res:-2 (No such file or directory) msg:"no global 41"
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.503030][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:0 seq:171 res:-22 (Invalid argument) msg:"unknown resource 30 op:7"
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.503058][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:29 seq:172 res:-2 (No such file or directory) msg:"no global 41"
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.503063][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:0 seq:174 res:-22 (Invalid argument) msg:"unknown resource 29 op:7"
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.503066][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:28 seq:175 res:-2 (No such file or directory) msg:"no global 41"
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.503094][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:0 seq:177 res:-22 (Invalid argument) msg:"unknown resource 28 op:7"
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.503100][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:27 seq:178 res:-2 (No such file or directory) msg:"no global 41"
aoû 04 14:34:55 desktop pipewire[2014]: [E][000000045.503103][private.h:241 pw_core_resource_errorv()] resource 0x15173f0: id:0 seq:180 res:-22 (Invalid argument) msg:"unknown resource 27 op:7"
```

zimbatm commented on nixos/modules/services/desktops/pipewire.nix:
```diff
@@ -25,17 +27,74 @@ in {
+    '';
+    environment.etc."alsa/conf.d/50-pipewire.conf".source = mkIf cfg.alsa.enable
+      "${pkgs.pipewire}/share/alsa/alsa.conf.d/50-pipewire.conf";
+    environment.sessionVariables.LD_LIBRARY_PATH =
+      (optional cfg.jack.enable "${pkgs.pipewire.lib}/lib/pipewire-0.3/jack") ++
+      (optional cfg.pulse.enable "${pkgs.pipewire.lib}/lib/pipewire-0.3/pulse");
```
Another way to do this is to link those into the systemPackages in the lib/ folder as glibc will automatically look for libraries there. This avoids the issue that users have to log out and back in for these changes to take effect.

nglen replied:
I tried to do this earlier, however it didn't seem to work. I believe that this is because programs manually specify dependencies on PulseAudio and JACK. Because of this they refer to the libraries by absolute path where they are successfully found before checking the default path. This means that programs end up using the original libraries instead of the PipeWire wrappers.

In other words:
`LD_LIBRARY_PATH` > hard-coded paths > /run/current-system/sw/lib

nglen commented:
> If I use `services.pipewire.enable = true;` without enabling `services.pipewire.alsa.enable = true;` I get:
> 
> ```
> error: The option `nodes.desktop.configuration.environment.etc.alsa/conf.d/50-pipewire.conf.source' is used but not defined.
> (use '--show-trace' to show detailed location information)
> ```

Thanks for noticing. I accidentally made a modification in refactoring which broke this without noticing. For files, `mkIf` needs to be placed on the attribute set and not on the source. I just pushed a fix.

nglen commented:
> can you give more info for how to test this? i was not able to get any kind of audio output with any of `aplay`, `paplay`, or `jack_simple_session` with this. `qjackctl` saw a jack server running, but `jack_simple_session` was not able to connect to it and neither was `cadence`. `pavucontrol` listed a number of devices in the "configuration" tab, but the only setting for any of them was "off".

I tested with the following among others:
- `mpv --ao=alsa <filename>`
- `mpv --ao=pulse <filename>`
- Chromium (PulseAudio, disable or set `LD_LIBRARY_PATH=""` to test ALSA)
- Firefox (PulseAudio and ALSA through `apulse`)
- `jack_simple_client` (JACK2)
- `alsamixer` (You may have to switch sound cards)
- `amixer` (You may have to manually specify your card with `-c`)
- `pavucontrol` (It doesn't show JACK programs only ALSA and PulseAudio)
- `ncpamixer` (same things as above)
- `qjackctl` (It shows everything but only names JACK programs)

I will say that the PulseAudio library in particular still has some unimplemented functions and weird oddities but this is on the PipeWire side. Be careful with `aplay` and `paplay` as I believe that they only support raw audio streams and I never could get to test them.

nglen commented:
@bbigras 

Having `hardware.pulseaudio.enable = false` should be correct, that's what I'm doing. Do you know whether the errors occur when the service starts or when some program is run. If the latter I could try the program on my end.

nglen commented:
I just realized where the problem is. I was using a slightly older version of nixpkgs than this pull request. Updating to this version breaks the daemon on my side. Reverting #93672 appears to fix the issues though I'm not sure why.

nglen commented:
Just found the issue in upstream `alsadatadir = join_paths('/usr', 'share', 'pulseaudio', 'alsa-mixer')`.

Since I can't reference GitLab, the commit is [this](https://github.com/PipeWire/pipewire/commit/1612f5e4d215bd5edf7d649d220b53ff1ed7c098).

I know this is the issue because manually symlinking the directory from `pulseaudio` "fixes" it.

ashkitten commented:
confirmed paplay and pavucontrol appear to work after reverting pipewire to 0.3.6, but i wasn't able to make qjackctl or cadence work, they couldn't find the jack server. didn't try with alsa.

ashkitten commented:
tried to build 0.3.9 but it gives this:
```
cycle detected in the references of '/nix/store/3r9gvywjzd8v671cb2bdh4a5afvbyllv-pipewire-0.3.9-lib' from '/nix/store/9y7fpwdr75f7fgka6afdmds82jif2sb9-pipewire-0.3.9'
```

bisected it down to https://github.com/PipeWire/pipewire/commit/154f981f97e8fe1abb471c6f3ef7d2a40f41b59f which is the same commit that changed `alsadatadir` from that absolute path with `/usr` to referencing `pipewire_datadir`. not sure what to do with that info...

nglen commented:
I just checked the paths, it looks like we should just need to move `$out/share/alsa-card-profile` to `$lib/share/alsa-card-profile`.

nglen commented:
On second thought, we might have to move the entirety of `$out/share` somehow since it's all referenced by the same variable.

ashkitten commented:
can we just change it to reference `pipewire_libdir` instead of `pipewire_datadir`? not sure if that's correct but it lets it build for me.

nglen commented:
I'm not sure either, but if it works it works.

ashkitten commented:
so here are all the things i've gotten working on my setup with `0.3.9`, in no particular order:
- `paplay`
- `pavucontrol`
- `qjackctl`
- `jack_simple_client`
- `aplay` (requires `-D pipewire`, it does not seem to set itself as the default alsa device)
- `firefox`

cadence still appears broken. can you test on your end?

nglen commented:
It's broken on my end as well.

nglen committed:
pipewire: 0.3.7 -> 0.3.9

Co-authored-by: Jan Tojnar <jtojnar@gmail.com>

jtojnar commented:
Changelogs look okay: https://github.com/PipeWire/pipewire/blob/0.3.9/NEWS#L1-L80

Ofono support was added, maybe we want to include it.

jtojnar commented on pkgs/development/libraries/pipewire/default.nix:
```diff
@@ -42,9 +42,11 @@ stdenv.mkDerivation rec {
+    sha256 = "0q781r32mnm3qy6xcdd2rnb8g50gdi7mi50zmdiq24s24sr8f8r9";
   };
 
+  patches = [ ./alsa-profiles-use-libdir.patch ];
```
```suggestion
  patches = [
    # Break up a dependency cycle between outputs.
    ./alsa-profiles-use-libdir.patch
  ];
```

jtojnar commented on pkgs/development/libraries/pipewire/alsa-profiles-use-libdir.patch:
```diff
@@ -0,0 +1,13 @@
+ spa_plugindir = join_paths(pipewire_libdir, spa_name)
+ 
+-alsadatadir = join_paths(pipewire_datadir, 'alsa-card-profile', 'mixer')
++alsadatadir = join_paths(pipewire_libdir, 'alsa-card-profile', 'mixer')
```
Can we just change the output instead of the whole path so it is less confusing. For example using:

```suggestion
+alsadatadir = join_paths(pipewire_libdir, '..', 'share', 'alsa-card-profile', 'mixer')
```

jtojnar commented on nixos/modules/services/desktops/pipewire.nix:
```diff
@@ -25,17 +27,75 @@ in {
+      pcm_type.pipewire {
+        libs.native = ${pkgs.pipewire.lib}/lib/alsa-lib/libasound_module_pcm_pipewire.so ;
+        ${optionalString enable32BitAlsaPlugins
+          "libs.32Bit = ${pkgs.pkgsi686Linux.pipewire.lib}/lib/alsa-lib/libasound_module_pcm_pipewire.so ;"}
```
Would it make sense to add the paths to `passthru` inside pipewire so that we could add tests ensuring the files exist like we did [in fwupd](https://github.com/NixOS/nixpkgs/blob/8e2b14aceb1d40c7e8b84c03a7c78955359872bb/pkgs/os-specific/linux/firmware/fwupd/default.nix#L279-L304)?

Full-fledged tests would be even better but that is even more work.

nglen replied:
Good idea, I just added some tests for this.

nglen committed:
pipewire: add testing

This adds two tests. One is for whether the paths used by the module are
present, while the other is for testing functionality of PipeWire
itself. This is done with the recent addition of installed tests by
upstream.

ashkitten commented:
if we can't change the fact that we're using `LD_LIBRARY_PATH` for this, should we at least make it point to some mutable location that the store paths will then be linked into, so that they can be updated without needing to modify the environment variable (and thus require logging back in)?

nglen commented:
That seems like a far better idea. If the path is module specific you would still need to re-log after enabling it but that's still an improvement. I guess the best way to do this would be to add new packages to `systemPackages` and then add a path like `/run/current-system/sw/lib/pipewire` to `LD_LIBRARY_PATH`.

ashkitten commented:
managed to get at least catia working by changing the `libjackso=` line in the cadence build with a path to the pipewire libjack.so (can't just override `libjack2 = pipewire.lib` because it's in a subdirectory. maybe we should provide a package that can be a slot-in replacement?). the rest of cadence seems to depend on jackdbus.

ashkitten commented:
tried running `jackdbus` but it gives `jackdbus: symbol lookup error: jackdbus: undefined symbol: jack_error_callback`.

that symbol is defined in pipewire's `libjack.so` so i tried adding that to `LD_PRELOAD`, which granted me `jackdbus: symbol lookup error: jackdbus: undefined symbol: jackctl_server_get_drivers_list`

this seems to be from jack2's `libjackserver.so` but is not defined in pipewire's version of that library, so i think i'm stuck unless pipewire decides to implement `jackdbus` or all of the requisite functions.

flokli commented on nixos/modules/services/desktops/pipewire.nix:
```diff
@@ -25,17 +35,76 @@ in {
+    environment.etc."alsa/conf.d/50-pipewire.conf" = mkIf cfg.alsa.enable {
+      source = "${pkgs.pipewire}/share/alsa/alsa.conf.d/50-pipewire.conf";
+    };
+    environment.sessionVariables.LD_LIBRARY_PATH = "/run/current-system/sw/lib/pipewire";
```
Is there some dispatcher library, to which `runpath` we could add this, similar to `addOpenGLRunpath` works with libglvnd?

nglen replied:
Not currently. How exactly would you use such a package? With OpenGL the programs depend on it directly while the programs we're dealing with depend on PulseAudio or JACK instead. We could create some kind of audio `runpath` and use it for PulseAudio and JACK as well but I'm not sure whether such a large change would be worth it, at least for now.

nglen replied:
I will note however, that if we did do this it would allow for better support of both 32bit programs on 64bit machines as well as sandboxed applications.

eadwu commented on nixos/modules/services/desktops/pipewire.nix:
```diff
@@ -5,8 +5,18 @@ with lib;
+  '';
+  pulse-libs = pkgs.runCommand "pulse-libs" {} ''
+    mkdir -p "$out/lib"
+    ln -s "${pkgs.pipewire.lib}/lib/pipewire-0.3/pulse" "$out/lib/pipewire"
```
`pipewire-*` might be better?

nglen replied:
You're probably right. The old method of using `LD_LIBRARY_PATH` directly required it to be fully qualified but that isn't necessary anymore.

eadwu commented on nixos/modules/services/desktops/pipewire.nix:
```diff
@@ -25,17 +35,76 @@ in {
-    environment.systemPackages = packages;
+    environment.systemPackages = [ pkgs.pipewire ]
+                                 ++ lib.optional cfg.jack.enable jack-libs
+                                 ++ lib.optional cfg.pulse.enable pulse-libs;
```
Since filenames can conflict with pulseaudio and jack, I believe these should be at a lower priority.

nglen replied:
Filenames shouldn't conflict since they are put in a `pipewire` subdirectory.

eadwu replied:
Ah sorry I read the derivation wrong.

Since you separated them, I thought you just put them in `$out/lib` instead of `$out/lib/pipewire`. Priorities and putting the libraries at `$out/lib` is one way to resolve the `$LD_LIBRARY_PATH` issue below though.

nglen replied:
The problem is that then the programs use the hard-coded rpath values before `systemPackages`. This means they find the original libraries instead of the replacements.

eadwu commented on nixos/modules/services/desktops/pipewire.nix:
```diff
@@ -25,17 +35,76 @@ in {
+      }
+    '';
+    environment.etc."alsa/conf.d/50-pipewire.conf" = mkIf cfg.alsa.enable {
+      source = "${pkgs.pipewire}/share/alsa/alsa.conf.d/50-pipewire.conf";
```
I'm not too sure but in my configuration I included this file with `<${pkgs.pipewire}/share/alsa/alsa.conf.d/50-pipewire.conf>` inside asoundrc.

nglen replied:
I believe this is the preferred way as it is what the JACK module uses.

https://github.com/NixOS/nixpkgs/blob/dd0a47e7ae2f6f33298c61f75aff2652e0a2f6d6/nixos/modules/services/audio/jack.nix#L237

nglen committed:
nixos/pipewire: use globs for the version

nglen committed:
pipewire: split pulse and jack emulation out

This splits PulseAudio and JACK emulation into separate outputs. Doing
so provides a number of benefits.

First it fixes pw-pulse and pw-jack. Prior to this they pointed to bogus
locations because the environment variables were not evaluated.
Technically fixing this only requires setting libpulse-path and
libjack-path to any absolute path not necessarily separate outputs but
it comes as a nice result.

Secondly it allows overriding libpulseaudio with pipewire.pulse in many
packages. This is possible because the new outputs have a more standard
layout.

nglen committed:
pipewire: remove direct dependency on PulseAudio

Since PulseAudio is required as a build input, the libraries mistakenly
reference the original libpulse.so rather than the custom one.

nglen committed:
nixos/pipewire: use provided udev rules

As of PipeWire 0.3.8, the library is bundled with a slightly modified
version of PulseAudio's udev rules.

jansol commented:
Is this blocked on something else or just waiting on progress of its own? Might be worth bumping the version to 0.3.11 while at it, as there seem to have been quite a few fixes for the JACK and pulseaudio modules.

On an unrelated note, is there any documentation for configuring pipewire? I tried pw-cat but it complains about no modules available, although pw-cli does show that it has detected all my audio interfaces.

nglen committed:
nixos/pipewire: add missing dbus dependency

Without this it is possible for PipeWire to fail in binding the dbus socket.

nglen committed:
pipewire: update to 0.3.11

nglen commented:
I was just finishing up some testing and bug fixing. I just pushed the changes I made so it should be ready. I also went ahead and bumped the version since you mentioned it. What's the best way to get this merged soon?

As far as I know, there aren't any good ways to directly configure PipeWire at the moment. [WirePlumber](https://pipewire.pages.freedesktop.org/wireplumber/daemon/configuration.html) should be used in the future which will hopefully improve this situation. If you need to mess with the `/etc/pipewire/pipewire.conf` file then you can set the `PIPEWIRE_CONFIG_FILE` variable. If you need to debug something `PIPEWIRE_DEBUG=4` should help. These can be set in nix using `systemd.user.services.pipewire.environment`.

bbigras commented:
services.pipewire.pulse and services.pipewire.jack don't exist anymore?

nglen commented:
What do you mean? `services.pipewire.pulse.enable` and `services.pipewire.jack.enable` should both exist.

nglen commented:
I split the PulseAudio and JACK libraries into separate outputs on the package for a number of reasons. The error is saying that these outputs don't exist. Are you using the versions of these packages provided in the newest commit or do you have some kind of overlay?

bbigras commented:
My bad, it seems I wasn't up-to-date on that part.

I'm trying pipewire with this:

```nix
  hardware.pulseaudio.enable = false;
  services.pipewire = {
    enable = true;
    pulse.enable = true;
    alsa.enable = true;
  };
```

and I don't see any output devices in `pavucontrol`.

```
❯ journalctl --user -b -u pipewire
-- Logs begin at Mon 2020-08-10 19:08:50 EDT, end at Fri 2020-09-11 11:21:41 EDT. --
sep 11 07:20:39 desktop systemd[1994]: Started Multimedia Service.
sep 11 07:20:39 desktop pipewire[2189]: [W][000000024.958623][module-rtkit.c:499 idle_func()] could not make thread realtime: No such file or directory
sep 11 11:20:55 desktop pipewire[2191]: [E][000000039.181170][bluez5-dbus.c:1436 get_managed_objects_reply()] GetManagedObjects() failed: org.freedesktop.DBus.Error.TimedOut
```


here's the log if I try to play something in vlc:

```
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.701642][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.713532][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.721438][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.727064][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.742179][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.750910][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.760771][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.769598][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.775374][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.780103][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2189]: [E][000000327.782795][private.h:241 pw_core_resource_errorv()] resource 0x15aadc0: id:32 seq:129 res:-2 (No such file or directory) msg:"no global 43"
sep 11 11:25:43 desktop pipewire[2189]: [E][000000327.783110][private.h:241 pw_core_resource_errorv()] resource 0x15aadc0: id:0 seq:131 res:-22 (Invalid argument) msg:"unknown resource 32 op:7"
sep 11 11:25:43 desktop pipewire[2189]: [E][000000327.783179][private.h:241 pw_core_resource_errorv()] resource 0x15aadc0: id:33 seq:132 res:-2 (No such file or directory) msg:"no global 43"
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.786492][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.789964][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.793158][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.797835][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.801226][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2189]: [E][000000327.804737][private.h:241 pw_core_resource_errorv()] resource 0x15aadc0: id:0 seq:134 res:-22 (Invalid argument) msg:"unknown resource 33 op:7"
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.805068][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.808834][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.812451][policy-node.c:549 rescan_node()] no node found for 43
sep 11 11:25:43 desktop pipewire[2191]: [W][000000327.817605][policy-node.c:549 rescan_node()] no node found for 43
[...]
```

I didn't use `apulse` at all for any of those commands.

nglen committed:
pipewire: remove old tests

nglen commented:
A few things to try in order:

- `systemctl --user restart pipewire`
- Disable unnecessary features:
  ```nix
  self: super: {
    pipewire = super.pipewire.override {
      ofonoSupport = false;
      nativeHspSupport = false;
    };
  }
  ```
- Downgrade PipeWire from 0.3.11:
  ```nix
  self: super: {
    pipewire = super.pipewire.overrideAttrs ({ ... }: rec {
      version = "0.3.9";

      src = super.fetchFromGitLab {
        domain = "gitlab.freedesktop.org";
        owner = "pipewire";
        repo = "pipewire";
        rev = version;
        sha256 = "0q781r32mnm3qy6xcdd2rnb8g50gdi7mi50zmdiq24s24sr8f8r9";
      };
    });
  }
  ```

Please tell me which, if any, of these fixes the problem so that I can try to hunt it down.

jansol commented:
Looking at the failing tests, the x86_64 looks weird, says all tests passed but still considers the build failed. There is also this line in the log: `'spa_pod_is_fraction(pod)' failed at ../spa/include/spa/pod/iter.h:315 spa_pod_get_fraction()`

The aarch64 tests on the other hand complain about not finding pulse and jack dirs, so maybe that is worth a look?  Or if pulse and jack are not supported on aarch64 to begin with, they should always be disabled there along with their tests.

nglen commented:
The tests were failing because they still checked the old location for the pulse/jack directories. The commit I just sent fixes this.

The `spa_pod_get_fraction` part confused me but it turns out that isn't a failing test but a debug message within a function which is expected to fail. Notice the assert [here](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/spa/tests/test-pod.c#L334) for a negative error code. Even more confusingly, the long object prints are completely unrelated debug information from somewhere else.

bbigras commented:
> Please tell me which, if any, of these fixes the problem so that I can try to hunt it down.

@nglen my bad. I was using your new module but I didn't override the pipewire pkg, so I was still on 0.3.7. :facepalm:  Sorry.

Also sorry about the delay. It took me a while to do many tests. I had some problems trying to use an overlay to override the pkg and my computer was running out of memory for some reason.

Now pipewire seems to work. I just tested with youtube on firefox and with vlc. The sound is really quiet with vlc. Even if the volume is 100%.

![Peek 14-09-2020 00-10](https://user-images.githubusercontent.com/24027/93043261-cb74f100-f640-11ea-8640-60b6e88b2689.gif)

The volume is also 100% in the player.

I didn't use apulse for that.

jansol commented:
I take it this is good to go then?

jchv commented:
When I try to run this, the pulse/jack compatibility does not seem to work as expected; maybe I am doing something wrong? If I `LD_PRELOAD` the compatibility libs manually, things work, but the `LD_LIBRARY_PATH` hack for some reason does not work for me. (I rebased the PR to master locally, that might have an impact.)

jansol commented:
Upstream updated to 0.3.12 with more pulse compatibility fixes, might help with vlc.

nglen commented:
@jchv Did you relog after first enabling the module? Environment variables are all set on login.

jchv commented:
> @jchv Did you relog after first enabling the module? Environment variables are all set on login.

I was rebooting after each rebuild. I'll give it another shot. (I also think the `LD_LIBRARY_PATH` stuff conflicts with SANE configuration too.)

nglen commented:
I'll look into trying it off master. How do you think that it conflicts with SANE though?

nglen commented:
It appears to still work for me after rebasing off of master so that shouldn't be the issue.

- Check that `LD_LIBRARY_PATH` points to the right directory `/run/current-system/sw/lib/pipewire` and that that directory both exists and contains the correct files.
- Try using `pw-pulse` and/or `pw-jack` and see if they work since they should do the same thing.
- Make sure that you not only enabled `services.pipewire.enable` but also `services.pipewire.pulse.enable` and/or `services.pipewire.jack.enable`.
- Also make sure that both `pulseaudio` and `jack` are disabled in your NIxOS config.

bbigras commented:
@nglen is it good enough that you could push the rebased commits to make it easier for us to test? I would just have to run `niv update pipewire` in my case.

jchv commented:
I'll take a look today and see if I still run into issues with SANE and the wrappers, however I am not going to be running Pipewire as a daily driver yet due to [pipewire/pipewire#297](https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/297). Still, I think I did have everything setup as you described.

jchv commented:
OK, here is the issue I run into with SANE:

```
building the system configuration...
error: The unique option `environment.sessionVariables.LD_LIBRARY_PATH' is defined multiple times, in:
 - /home/john/Sources/nixpkgs/nixos/modules/services/hardware/sane.nix
 - /home/john/Sources/nixpkgs/nixos/modules/services/desktops/pipewire.nix.
(use '--show-trace' to show detailed location information)
```

nglen committed:
nixos/pipewire: fix variable merging

While both strings and lists are valid, lists are the more accurate type
and only they can properly merge with each other.

nglen commented:
@jchv Thanks for noticing that. I accidentally used a string instead of a list, both are valid values but only lists merge. The commit I just pushed should fix it.

jchv commented:
Thanks, that worked like a charm. Now onto the next problem... the wrappers still don't work. I can't help but feel the problem for me must be PEBCAK.

From within my GNOME/Wayland session:

```
$ echo $LD_LIBRARY_PATH 

```

...oddly, `LD_LIBRARY_PATH` is empty?

I checked to make sure I wasn't doing anything weird, but nothing that's currently active in my configuration twiddles with `LD_LIBRARY_PATH` or even `sessionVariables`. Hmm. Looking into how to debug this. (I did reboot already.)

jchv commented:
One last note. I tried applying the patch in https://github.com/NixOS/nixpkgs/issues/90201#issuecomment-682373903 and now everything works largely as expected. So for future readers, if you want to use this on GNOME 3 + Wayland, you will probably need to patch this for now. It would be nice if there could be a generalized workaround that just finds a way to pipe `LD_LIBRARY_PATH` through explicitly so we can at least unbreak stuff... maybe I can try a couple of things and see if anything is possible.

With all that being said, this is not too related to this PR specifically although I have a feeling it will have an impact on it for sure.

bbigras commented:
> Upstream updated to 0.3.12 with more pulse compatibility fixes, might help with vlc.

@jansol thanks! I tried pipewire 0.3.12 with 59cbf9746c162359b9bbad334be8b43a2a1548aa and vlc's volume is still super quiet.

jansol commented:
Sooo.... wait for next week's (presumably) 0.3.14 and then merge unless there are big regressions?

Having the nixos module merged would make further testing easier anyway...

worldofpeace commented:
I would only like to see this merged after us from @NixOS/gnome approve on this. The service is default enabled in GNOME and we maintain it. (will should change the maintainers to `teams.gnome`)

jchv commented:
As far as I can tell this will not work in GNOME+Wayland at all, because of #90201. Am I missing something?

worldofpeace commented:
> As far as I can tell this will not work in GNOME+Wayland at all, because of #90201. Am I missing something?
> 
> edit: To be clear I still eagerly await this PR to be merged, I just also suspect it may be unwise to make it a default in GNOME until that problem can be worked around.

Yeah, I believe GNOME is only at the phase of "try this feature out" and not "let's use this now".
We're actually going to do https://github.com/NixOS/nixpkgs/pull/99147#pullrequestreview-499715152, and turn it off by default. So #90201 won't be an issue in the default configuration.

ashkitten commented:
how can we make this work with fhsenv applications like steam and lutris? steam games work fine with `pw-pulse %command%` as the launch args, but not by default.

jtojnar reviewed (commented):
Only few edges to polish but I do not foresee any compatibility issues with GNOME, as long as the shims are off by default. GNOME team can work on fixing the conflict with disabled-by-default rt-scheduling (#90201) later.

I also checked the changelogs up to 0.3.13 and we will want to update for the crash fixes but other than that, I did not see anything interesting. Well, 0.3.11 introduced rudimentary hsphfpd (packaging status tracked in #87910) but according to pali’s comments in [pipewire/pipewire!295](https://gitlab.freedesktop.org/pipewire/pipewire/merge_requests/295) it is not ready to be enabled by default yet.

L-as commented:
FWIW I've been using this and noticed that switching from pulseaudio to pipewire with nixos-rebuild switch doesn't work at all.

jtojnar commented:
@L-as You need to log-in again as you cannot modify the environment of running session.

nglen committed:
pipewire: update to 0.3.12

nglen committed:
nixos/pipewire: cleanup module

nglen committed:
pipewire: use meson options for installed_tests

nglen committed:
pipewire: remove pulseaudio references

This replaces the temporary RPATH fix with the new upstream patch  from
jtojnar. It also removes the reference to the dev output.

nglen commented on nixos/modules/services/desktops/pipewire.nix:
```diff
@@ -25,17 +35,78 @@ in {
+          description = ''
+            Enable transparent PulseAudio emulation using LD_LIBRARY_PATH.
+          '';
+        };
```
Done

nglen commented on nixos/modules/services/desktops/pipewire.nix:
```diff
@@ -25,17 +35,78 @@ in {
+    environment.etc."alsa/conf.d/50-pipewire.conf" = mkIf cfg.alsa.enable {
+      source = "${pkgs.pipewire}/share/alsa/alsa.conf.d/50-pipewire.conf";
+    };
+    environment.sessionVariables.LD_LIBRARY_PATH = [ "/run/current-system/sw/lib/pipewire" ];
```
Good idea, done.

jtojnar commented on pkgs/development/libraries/pipewire/default.nix:
```diff
@@ -33,18 +37,37 @@ let
+  ];
+
+  postPatch = ''
+    substituteInPlace meson.build --subst-var-by installed_tests_dir "$installedTests"
```
I cherry-picked these into GNOME branch with the suggestion applied:

https://github.com/NixOS/nixpkgs/commit/59445069e60746af7939911d58137c3e8ce649ae
https://github.com/NixOS/nixpkgs/commit/8c21e43c9c22a74bcb735483b0906127274f2724
https://github.com/NixOS/nixpkgs/commit/c838006f6076d770170b3cce2db25eb78ee0ce35

nglen replied:
This is a far better idea. I just implemented it.

nglen commented on nixos/modules/services/desktops/pipewire.nix:
```diff
@@ -5,8 +5,18 @@ with lib;
+                           && pkgs.stdenv.isx86_64
+                           && pkgs.pkgsi686Linux.pipewire != null;
 
+  jack-libs = pkgs.runCommand "jack-libs" {} ''
+    mkdir -p "$out/lib"
+    ln -s "${pkgs.pipewire.jack}/lib" "$out/lib/pipewire"
+  '';
+  pulse-libs = pkgs.runCommand "pulse-libs" {} ''
+    mkdir -p "$out/lib"
+    ln -s "${pkgs.pipewire.pulse}/lib" "$out/lib/pipewire"
+  '';
```
There are two reasons I didn't do that. One is that I don't like the package output should depend on module internals. That seems like a potential backwards compatibility issue in the future. We probably want to change the module eventually to avoid using `LD_LIBRARY_PATH` once a way is setup (and the inevitable mass recompilation acceptable). The second is that currently the output can be used as an override on FHS and wrapper style packages, the exact ones that tend to cause problems with this setup. For example the following works:

  ```nix
  pkgs.tor-browser-bundle-bin.override { libpulseaudio = pkgs.pipewire.pulse; }
  ```

This is, of course, only valid when development information like headers aren't required.

nglen commented on pkgs/development/libraries/pipewire/alsa-profiles-use-libdir.patch:
```diff
@@ -0,0 +1,13 @@
+ spa_plugindir = join_paths(pipewire_libdir, spa_name)
+ 
+-alsadatadir = join_paths(pipewire_datadir, 'alsa-card-profile', 'mixer')
++alsadatadir = join_paths(pipewire_libdir, '..', 'share', 'alsa-card-profile', 'mixer')
```
This directory is actually used by the SPA plugin and not the compatibility layer. The reason it was moved to `pipewire_libdir` is that there is a direct dependency link to which was causing cyclic dependencies. It might make more since to move it to `$lib/lib/spa-0.2/alsa/mixer` to make it more obvious.

The compatibility layer directories are `$lib/lib/alsa-lib` and `$out/share/alsa`. These could be moved to a separate output but are quite small and wouldn't remove the dependency on `alsaLib` due to the previously mentioned SPA plugin. I am open to this if it is preferred though.

jtojnar replied:
I guess this is fine for now, thanks for the explanation.

nglen commented on pkgs/development/libraries/pipewire/default.nix:
```diff
@@ -78,12 +101,40 @@ stdenv.mkDerivation rec {
   doCheck = true;
 
+  preFixup = ''
+    # The rpaths mistakenly points to libpulseaudio instead
```
Fixed

nglen commented on pkgs/development/libraries/pipewire/default.nix:
```diff
@@ -78,12 +101,40 @@ stdenv.mkDerivation rec {
+    # The rpaths mistakenly points to libpulseaudio instead
+    for file in "$pulse"/lib/*.so; do
+      oldrpath="$(patchelf --print-rpath "$file")"
+      patchelf --set-rpath "$pulse/lib:$oldrpath" "$file"
```
Thanks, I just converted it to run off of that instead. The patch can then be removed once the next update rolls out.

nglen commented on nixos/modules/services/desktops/pipewire.nix:
```diff
@@ -25,17 +35,78 @@ in {
 
     systemd.user.sockets.pipewire.wantedBy = lib.mkIf cfg.socketActivation [ "sockets.target" ];
-  };
+    systemd.user.services.pipewire.bindsTo = [ "dbus.service" ];
```
Added

jtojnar commented:
Why not update directly to 0.3.13? Otherwise looks good to me.

nglen committed:
pipewire: cleanup path testing

nglen committed:
pipewire: update to 0.3.13

nglen commented:
Sorry, that was a mistake on my part. Just updated.

jtojnar commented:
Verified that screen recording works in GNOME Shell and `pipewire.tests` succeeds.

Thanks.

bbigras commented:
Thanks everybody! :tada:
